[15.0][MIG] web_form_banner: Migration to 15.0

AungKoKoLin1997 opened the pull request:
Based on https://github.com/OCA/web/pull/3286

The module adds configurable banners for backend form views. Define rules per model
(and optionally per view) to show context-aware alerts with a chosen severity (info/warning/danger).

Messages can be plain text with ${placeholders} or fully custom HTML; visibility,
severity, and values are computed server-side via a safe Python expression.

Banners are injected just before or after a target node (default: //sheet) and refresh
on form load/save/reload.

@qrtl QT5747

yostashiro committed:
[ADD] web_form_banner

AungKoKoLin1997 committed:
[IMP] web_form_banner: black, isort, prettier

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 15.0

Refactor JS:
- Switch to the ESM style
- Change var to const/let
- Remove unused parts (diff, el.dataset.wfbTriggerFields)

Miscellaneous improvements:
- list view optional
- Switch archive box icon to banner

Fix _render_html for Odoo 15:
- In Odoo 12, html_escape returned a plain string, so
  replacing newlines with <br/> worked as expected.
- In Odoo 15, html_escape uses markupsafe.escape, which
  escapes inserted <br/> tags resulting in &lt;br/&gt; in the output.
- Updated _render_html to escape each line individually
  and join with literal <br/> tags to match Odoo 12 behavior
  while remaining safe for HTML rendering.

Co-authored-by: Yoshi Tashiro <tashiro@quartile.co>

yostashiro commented on web_form_banner/static/src/js/web_form_banner.esm.js:
```diff
@@ -0,0 +1,216 @@
+    var first = function () {
+        for (var i = 0; i < arguments.length; i++) {
+            var v = arguments[i];
+            if (v !== null && v !== "") return v;
```
Do you have a reason of this change? I think this change should be reverted to be safe, otherwise.
```suggestion
            if (v != null && v !== "") return v;
```

yostashiro commented on web_form_banner/static/src/js/web_form_banner.esm.js:
```diff
@@ -0,0 +1,216 @@
+            var k = kv[0],
+                v = kv[1],
+                t = typeof v;
+            if (v === null || t === "string" || t === "number" || t === "boolean") {
```
This change should be reverted.

```suggestion
            if (v == null || t === "string" || t === "number" || t === "boolean") {
```

yostashiro commented on web_form_banner/static/src/js/web_form_banner.esm.js:
```diff
@@ -0,0 +1,216 @@
+odoo.define("web_form_banner.save_plus_load", function (require) {
```
We should switch the content to the ESM style.

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -0,0 +1,287 @@
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools import html_escape, safe_eval
```
To avoid unnecessary diffs. Use `tools.safe_eval` for `time` and `datetime` in `_base_eval_ctx_static()`.
```suggestion
from odoo import _, api, fields, models, tools
from odoo.exceptions import ValidationError
from odoo.tools import html_escape
from odoo.tools.safe_eval import safe_eval
```

yostashiro commented:
@AungKoKoLin1997 Please confirm my commits and sort out the commit history. I think my changes can be squashed into the migration commit but comment the changed points.

yostashiro commented:
@AungKoKoLin1997 Can you also please switch the archive box icon to the banner?

<img width="283" height="144" alt="image" src="https://github.com/user-attachments/assets/0c56ed34-d475-4ffc-9046-cc818ef9febf" />

yostashiro reviewed (approved):
Code review and functional test.

Sayaka-qrtl commented on web_form_banner/README.rst:
```diff
@@ -0,0 +1,271 @@
+Usage
+=====
+
+#. Go to *Settings > Tachnical > User Interface > Form Banner Rules* and create a rule.
```
Typo: "Tachnical" → "Technical"

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -0,0 +1,290 @@
+    @api.model
+    def compute_message(self, rule_id, model, res_id, form_vals=None):
+        """Return {visible, severity, html} for the given rule and record."""
+        logging.info(form_vals)
```
We should either remove this or use _logger.debug(). The log gets too noisy, otherwise.

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -0,0 +1,290 @@
+        if not rule.exists() or not rule.active:
+            return {"visible": False}
+        values = self._sanitize_values(model, form_vals)
+        logging.info(values)
```
Same as above.

yostashiro reviewed (commented):
@AungKoKoLin1997 Please go ahead and squash commits.

Sayaka-qrtl reviewed (approved):
Functional test. LGTM

yostashiro commented:
@AungKoKoLin1997 Just noticed an issue with banner presentation. Can you please follow up on this?

<img width="912" height="490" alt="image" src="https://github.com/user-attachments/assets/adc0a6ce-10b9-426f-aab8-de392566fcb6" />

AungKoKoLin1997 commented:
> Just noticed an issue with banner presentation. Can you please follow up on this?

@yostashiro Fixed it and explain the changes in the migration commit message as well.

yostashiro reviewed (approved):
Thanks for the fix!

AungKoKoLin1997 committed:
[IMP] web_form_banner: Partially support in-group target XPath

- Add (partial) support for elements inside groups to be a target
  XPath without distorting the presentation
- Remove obsolete bits
- Remove lru_cache decorator

Co-authored-by: Yoshi Tashiro <tashiro@quartile.co>

AungKoKoLin1997 commented:
@yostashiro I added some changes that are to make consistent with odoo 16 changes.
Please review my last two commits.

nobuQuartile reviewed (approved):
LGTM
Functional review

AungKoKoLin1997 committed:
[IMP] web_form_banner: add summary and maintainers to manifest
